sources/host-ctr/pidnamespaces/pid_namespaces.go

```diff
@@ -0,0 +1,165 @@
+	manifestPath string = "/etc/host-containers/info/pid-namespaces.json"
+)
+
+// HostCtrPidNs represents the list of PID namespaces
+type HostCtrPidNs struct {
+	PidNamespaces []int64 `json:"pid-namespaces,omitempty"`
+}
+
+// GetTaskPidNs returns the PID NS associated with the container task PID or error on failure
+func GetTaskPidNs(pid uint32) (int64, error) {
+	pidNsFile, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+	if err != nil {
+		return -1, fmt.Errorf("failed to open %d PID namespace file", pid)
+	}
+	pidNsFd := int(pidNsFile.Fd())
+
+	// Get file descriptor that refers to a parent namespace, ioctl() request '_IO(NSIO, 0x2)'
+	_, err = unix.IoctlGetInt(pidNsFd, unix.NS_GET_PARENT)
+	if err == unix.EPERM {
+		// If we can't access the parent PID namespace, this means the current container task is sharing the root PID
+		// namespace. We don't want to track the root PID namespace in the manifest, so we return -1.
+		return -1, nil
+	} else if err != nil {
+		return -1, errors.Wrap(err, fmt.Sprintf("ioctl failed to 'NS_GET_PARENTNS' for '%s'", pidNsFile.Name()))
+	}
+
+	// The PID namespace is represented by the inode number of the task's PID namespace file
+	var sb unix.Stat_t
+	if err := unix.Fstat(pidNsFd, &sb); err != nil {
+		return -1, errors.Wrap(err, "failed to stat task PID namespace fd")
+	}
+
+	return int64(sb.Ino), nil
+}
+
+// AddPidNsToManifest takes a given PID's namespace and adds it to the internal manifest
+func AddPidNsToManifest(pidNs int64) (err error) {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	ok, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil && !ok {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+
+	// Add task PID NS to manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	tmpPidNsList := &HostCtrPidNs{}
+	hostCtrPidNsList := &HostCtrPidNs{}
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &tmpPidNsList)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal PID NS manifest")
+		}
+		for _, ele := range tmpPidNsList.PidNamespaces {
+			// If it already exists in the manifest list for whatever reason, we can just return without doing anything.
+			if ele == pidNs {
+				return nil
+			}
+		}
+	}
+	hostCtrPidNsList.PidNamespaces = append(tmpPidNsList.PidNamespaces, pidNs)
+	if err = writeManifest(hostCtrPidNsList, manifest); err != nil {
+		return err
+	}
+	if err = fileLock.Unlock(); err != nil {
```

Comment:
Hmm on second thought - will this keep the file lock if we return earl on an error? I guess the benefit of having the `defer` is that it will always call that 😢 how often do we expect a `Syscall` to fail deep in the flock code ...? Maybe go back to throwing away the error if that's the easiest thing and prevents a deadlock

Reply:
Yeah, that's what I was thinking as well. I can return an error from a deferred function, so I'll just do that again.

Reply:
Also to answer 
>  how often do we expect a Syscall to fail deep in the flock code

I do not expect this to ever error if the previous call to obtain the lock has succeeded. Of all the possible errors that can be returned by `flock(2)`, only 1 of them can happen during unlock; it only errors if the underlying file descriptor becomes invalid (e.g. lock file deleted or something)

Reply:
> using named error vars

Excellent solution!

```go
defer func() {
	if err = fileLock.Unlock(); err != nil {
		err = errors.Wrap(err, "failed to release lock for writing to PID NS manifest")
	}
}()
```